### PR TITLE
Update user validation

### PR DESCRIPTION
**Pull request by davidrapson**
Updates the user authentication routes to use `joi` for validation. There's a fair bit going on here, so will add inline comments to expand on the changes here.

### CODE REVIEW

File: controllers/user/forgotten-password.js [hunk 1]
```diff
@@ -44,37 +38,51 @@ async function processResetRequest(req, res, user) {
...
 }
 
 router
     .route('/')
     .all(requireUnauthed)
-    .get(renderRequestReset)
-    .post(validators.emailAddress, handleRequestReset);
+    .get(renderForm)
+    .post(async function(req, res) {
+        const validationResult = Joi.object({
```
Comment (davidrapson) [inline controllers/user/forgotten-password.js]: Validation here is only to check that the username is a valid format (e.g. an email)

File: controllers/user/login.js [hunk 2]
```diff
@@ -35,13 +37,11 @@ router
...
         // @TODO consider rate limiting?
-        passport.authenticate('local', (err, user, info) => {
-
+        passport.authenticate('local', (err, user) => {
             if (!user) {
                 // User not valid, send them to login again
-                res.locals.errors = [{ msg: info.message }];
-                res.locals.formValues = req.body;
-                return renderForm(req, res);
+                const errors = [{ msg: 'Username or password is invalid' }];
```
Comment (davidrapson) [inline controllers/user/login.js]: Noticed whilst testing this that the errors returned to the front-end was the raw passport error. For login we only want to render a generic error that the log in credentials weren't valid. There was no express-validator code on this route before, we may want to add some basic validation like the forgotten password screen, if only to check that the username is an email address.

File: controllers/user/reset-password.js [hunk 3]
```diff
@@ -56,50 +49,72 @@ router
...
 
-        if (!changePasswordToken) {
-            res.redirect('/user/login');
+        if (!token) {
+            redirect(req, res);
         } else {
-            // store the token to pass on to other requests (eg. outside the URL)
-            res.locals.token = changePasswordToken;
-            const errors = validationResult(req);
+            try {
```
Comment (davidrapson) [inline controllers/user/reset-password.js]: This controller has the biggest structural changes:

- Separates out verification of the token from checking the user
- Decodes the token first and then validates the new password, passes in the username from the token to allow us to check against conditional rules e.g. password must not match username
- Only once we know the format of the data is good do we lookup the user and try to update the password

File: controllers/user/views/forgotten-password.njk [hunk 4]
```diff
@@ -21,18 +21,22 @@
...
                     name: 'username',
                     label: 'Email address',
+                    explanation: 'Enter the email address associated with your account.',
                     isRequired: true,
                     value: formValues['username']
                 }, errors = errors) }}
 
                 <div class="form-actions">
-                    <input type="submit" class="btn" value="Reset password request" />
+                    <input type="submit" class="btn" value="Request password reset" />
```
Comment (mattandrews) [inline controllers/user/views/forgotten-password.njk]: 🙃

Comment (davidrapson) [reply controllers/user/views/forgotten-password.njk]: Honestly not sure If this more or less grammatically correct now. We might be better off with simply "Reset password:

**Pull request merged**